src/python/grpcio/grpc/_cython/_cygrpc/grpc_gevent.pyx.pxi

```diff
@@ -14,418 +14,91 @@
-    # Avoid running any Python code after setting the exception
-    cpython.PyErr_SetObject(exc_info[0], exc_info[1])
-    return GRPC_ERROR_CANCELLED
-
-###############################
-### Initializer ###############
-###############################
+import atexit
+import errno
+import sys
 
-cdef grpc_socket_vtable gevent_socket_vtable
-cdef grpc_custom_resolver_vtable gevent_resolver_vtable
-cdef grpc_custom_timer_vtable gevent_timer_vtable
-cdef grpc_custom_poller_vtable gevent_pollset_vtable
+gevent_hub = None
+g_gevent_pool = None
+g_gevent_threadpool = None
+g_gevent_activated = False
+
+
+cdef queue[void*] g_greenlets_to_run
+cdef condition_variable g_greenlets_cv
+cdef mutex g_greenlets_mu
+cdef bint g_shutdown_greenlets_to_run_queue = False
+
+
+cdef _submit_to_greenlet_queue(object cb, tuple args):
+  cdef tuple to_call = (cb,) + args
+  cdef unique_lock[mutex]* lk
+  Py_INCREF(to_call)
+  with nogil:
+    lk = new unique_lock[mutex](g_greenlets_mu)
+    g_greenlets_to_run.push(<void*>(to_call))
+    del lk
+    g_greenlets_cv.notify_all()
+
+
+cdef object await_next_greenlet():
+  cdef unique_lock[mutex]* lk
+  with nogil:
+    # Cython doesn't allow us to do proper stack allocations, so we can't take
+    # advantage of RAII.
+    lk = new unique_lock[mutex](g_greenlets_mu)
+    while not g_shutdown_greenlets_to_run_queue:
+      if not g_greenlets_to_run.empty():
+        break
+      g_greenlets_cv.wait(dereference(lk))
+  cdef object to_call = <object>g_greenlets_to_run.front()
+  Py_DECREF(to_call)
+  g_greenlets_to_run.pop()
+  del lk
+  return to_call
+
+def spawn_greenlets():
+  while True:
+    to_call = g_gevent_threadpool.apply(await_next_greenlet, ())
+    fn = to_call[0]
+    args = to_call[1:]
+    fn(*args)
+
+def shutdown_await_next_greenlet():
+  global g_shutdown_greenlets_to_run_queue
+  cdef unique_lock[mutex]* lk
+  with nogil:
+    lk = new unique_lock[mutex](g_greenlets_mu)
+    g_shutdown_greenlets_to_run_queue = True
+  del lk
+  g_greenlets_cv.notify_all()
 
 def init_grpc_gevent():
   # Lazily import gevent
-  global gevent_socket
-  global gevent_g
   global gevent_hub
-  global gevent_event
-  global g_event
-  global g_pool
+  global g_gevent_threadpool
+  global g_gevent_activated
+  global g_interrupt_check_period_ms
+  global g_gevent_pool
+
   import gevent
-  gevent_g = gevent
-  import gevent.socket
-  gevent_socket = gevent.socket
-  import gevent.hub
-  gevent_hub = gevent.hub
-  import gevent.event
-  gevent_event = gevent.event
   import gevent.pool
 
-  g_event = gevent.event.Event()
-  g_pool = gevent.pool.Group()
-
-  def cb_func(cb, args):
-    _spawn_greenlet(cb, *args)
-  set_async_callback_func(cb_func)
+  gevent_hub = gevent.hub
+  g_gevent_threadpool = gevent_hub.get_hub().threadpool
 
-  gevent_resolver_vtable.resolve = socket_resolve
-  gevent_resolver_vtable.resolve_async = socket_resolve_async
+  g_gevent_activated = True
+  g_interrupt_check_period_ms = 2000
```

Comment:
Since completion queue polling no longer happens on the main thread, we no longer need to pause completion queue polling to check for signals. This is actually _more_ responsive.